README.md
# WarCry

This ESO Addon is mainly used by my Raidgroup but can also be interesting for others that want to have some fun with their team


https://www.esoui.com/downloads/info3191-WarCry.html


The main purpose is to always be able to playback the WarCry emote. It is the one you get from the open dungeon in wrothgar.
The leader of a group can use a predefined word in the /party chat to trigger the collectible on all group members. And it looks awesome

It is based on the LibWarCry Library that i created and also uses it as a dependency. It can be found here: https://github.com/m00nyONE/LibWarCry and here: https://www.esoui.com/downloads/info3212-LibWarCry.html

CHANGELOG.md
# 1.2.0
- add new WarCrys and an experimental outfit swap
- new: bathtime, unchained animals, bubbletime

# 1.1.1
- api Bump

# 1.1.0
- add some new emotes

# 1.0.4
- fixed that i missed the unregisterforevent after addon is loaded

# 1.0.3
- adding some user requested warcrys ( cherry, enlight me, army, firebreath)

# 1.0.2
- api bump
- fix title being off by one character

# 1.0.1
- api bump & better documentation in README.md

# 1.0.0
- moving code to LibWarCry Library
- adding LibWaCry as dependency

# 0.3.1
- hotfixing bugs & more debugging

# 0.3.0
- reworking the whole base - now the listening string is the identifier for the WarCry
- adding "roar"
- convert addon to a lib - will be finished in the next update

# 0.2.0
- adding group sync via chat. only the leader of the group is allowed to trigger the animation for the group members
- adding ability to disable the chat listener

# 0.1.0
- first version
- adding keybinds for the emote to trigger it whenever you want

Menu.lua
local LAM2 = LibAddonMenu2

local panelData = {
    type = "panel",
    name = "WarCry",
    displayName = "WarCry",
    author = "|cFFC0CBm00ny|r",
    version = "0.3.1",
    website = "https://www.esoui.com/downloads/fileinfo.php?id=3191",
    slashCommand = "/wcsettings",	--(optional) will register a keybind to open to this panel
    registerForRefresh = true,	--boolean (optional) (will refresh all options controls when a setting is changed and when the panel is shown)
    registerForDefaults = true,	--boolean (optional) (will set all options controls back to default values)
}

local optionsTable = {
    [1] = {
        type = "header",
        name = "group control settings",
        width = "full",	--or "half" (optional)
    },
    [2] = {
        type = "checkbox",
        name = "synced WarCry",
        tooltip = "enables/disables listening for chat events to sync warcry in the group",
        getFunc = function() return WarCry.savedVariables.groupCommands end,
        setFunc = function(value) WarCry.toggleListener(value) end,
        width = "half",	--or "half" (optional)
        default = true,
        --warning = "Will need to reload the UI.",	--(optional)
    },
    [3] = {
        type = "texture",
        --image = "EsoUI\\Art\\ActionBar\\abilityframe64_up.dds",
        image = "/esoui/art/icons/trophy_malacaths_wrathful_flame.dds", -- needs to be dynamic
        imageWidth = 64,	--max of 250 for half width, 510 for full
        imageHeight = 64,	--max of 100
        width = "half",	--or "half" (optional)
    },
}

LAM2:RegisterAddonPanel("WarCryMenu", panelData)
LAM2:RegisterOptionControls("WarCryMenu", optionsTable)

WarCry_API.lua
WarCry = WarCry or {}
WarCry.name = "WarCry"
WarCry.color = "8B0000"
WarCry.credits = "@m00nyONE"
WarCry.slashCmd = "/wc"
WarCry.variableVersion = 1
WarCry.List = {}
WarCry.defaultVariables = {
    groupCommands = true,
}

function WarCry:CreateWarCry(name, ids)
    if name == nil then d("|c8B0000WarCry:|r name must not be nil!") return end
    if type(ids) ~= "table" then d("|c8B0000WarCry:|r ids must be a table!") return end

    for index, value in ipairs(ids) do
        if type(value) ~="number" then d("|c8B0000WarCry:|r id must be a number!") return end
    end

    ---table.insert(WarCry.List, name, ids)
    WarCry.List[name] = {}
    WarCry.List[name].collectibleIDs = ids

    d("|c8B0000WarCry:|r " .. name .. " has been created")
end

function WarCry.PlayWarCry(name)
    if name == nil then return end
    if type(name) ~= "string" then return end

    local playable = {}
    if WarCry.List[name] ~= nil then
        for index, value in ipairs(WarCry.List[name].collectibleIDs) do
            if GetCollectibleUnlockStateById(value) == 2 then
                table.insert(playable, value)
            else 
                d("You are missing |H1:collectible:".. value .."|h|h")
            end
        end

        if #playable == 0 then  return end
        --- play random collectible from array
        UseCollectible(playable[ math.random(#playable)])
        return
    end

end

WarCry.chatCallback = function(_, messageType, _, message, _, fromDisplayName)
    -- check if message is in group chat
    if messageType == CHAT_CHANNEL_PARTY then
        -- check if message is the one we are listening for
        if WarCry.List[string.lower(message)] ~= nil then
            if fromDisplayName == GetUnitDisplayName(GetGroupLeaderUnitTag()) then
                -- play the warcry from the list
                WarCry.PlayWarCry(string.lower(message))
            end
        end
    end
end

function WarCry.toggleListener(value)
    -- toggle listener
    if value then
       WarCry.startChatListener()
    else
        WarCry.stopChatListener()
    end

    -- write value to saved vars
    WarCry.savedVariables.groupCommands = value
end

function WarCry.startChatListener()
    EVENT_MANAGER:RegisterForEvent(WarCry.name, EVENT_CHAT_MESSAGE_CHANNEL, WarCry.chatCallback)
    d("|c8B0000WarCry: group listener enabled|r")
end
function WarCry.stopChatListener()
    EVENT_MANAGER:UnregisterForEvent(WarCry.name, EVENT_CHAT_MESSAGE_CHANNEL)
    d("|c8B0000WarCry: group listener disabled|r")
end

--- create SLASH_COMMAND that can play every warcry available in the list
SLASH_COMMANDS["/warcry"] = function(str)
    if str ~= nil then
        if WarCry.List[string.lower(str)] ~= nil then
            WarCry.PlayWarCry(string.lower(str))
        end
    end
end

function WarCry.loadDefaults()
    local ahuListenString = "ahu"
    local ahuCollectibleIDs = {353}
    local roarListenString = "roar"
    local roarCollectibleIDs = {9432, 9590}

    WarCry:CreateWarCry(ahuListenString, ahuCollectibleIDs)
    WarCry:CreateWarCry(roarListenString, roarCollectibleIDs)
end

function WarCry.OnAddOnLoaded(event, addonName)
    if addonName == WarCry.name then
        --EVENT_MANAGER:RegisterForEvent("WarCry.name", EVENT_PLAYER_ACTIVATED, self.Init)

        WarCry.savedVariables = WarCry.savedVariables or {}
        WarCry.savedVariables = ZO_SavedVars:NewAccountWide("WarCryVars", WarCry.variableVersion, nil, WarCry.defaultVariables, GetWorldName())

        WarCry.loadDefaults()

        if WarCry.savedVariables.groupCommands == true then
            WarCry.startChatListener()
        end
    end
end

EVENT_MANAGER:RegisterForEvent(WarCry.name, EVENT_ADD_ON_LOADED, WarCry.OnAddOnLoaded)

Data.lua
WarCry = WarCry or {}
WarCry.List = {
    ["rake"] = {"/rake"},

    ["dance"] = {"/dance"},
    ["dancerandom"] = {
        "/dancebreton",
        "/dancedunmer",
        "/dancealtmer",
        "/dancekhajiit",
        "/dancenord",
        "/danceredguard",
        "/danceorc",
        "/dancebosmer",
        "/danceimperial",
        "/dancedrunk",
    },
    ["danceredguard"] = {"/danceredguard"},
    ["dancebreton"] = {"/dancebreton"},
    ["dancedunmer"] = {"/dancedunmer"},
    ["dancealtmer"] = {"/dancealtmer"},
    ["dancekhajiit"] = {"/dancekhajiit"},
    ["dancenord"] = {"/dancenord"},
    ["danceorc"] = {"/danceorc"},
    ["dancebosmer"] = {"/dancebosmer"},
    ["danceimperial"] = {"/danceimperial"},
    ["dancedrunk"] = {"/dancedrunk"},

    ["lute"] = {"/lute"},
    ["flex"] = {"/flex"},
    ["bullhorns"] = {"/bullhorns"},
    ["bucketsplash"] = {"/bucketsplash"},

    ["ahu"] = {353},
    ["roar"] = {9432, 9590},
    ["bathtime"] = {753}, -- equips everyone with the nord bath costume

    -- user requests
    ["cherry"] = {1108},
    ["enlight me"] = {336},
    ["army"] = {350},
    ["firebreath"] = {600},
    ["unchained animals"] = {5242},
    ["bubbletime"] = {10248}
}